### PR TITLE
docs: WIP ADR about transaction aggregates

### PR DESCRIPTION
**Pull request by iloveagent57**
Still very open to feedback, but here's a proposed approach to help us de-couple runtime dependencies in the ``can_redeem`` and ``redeem`` flows, and in the domain of groups.

### CODE REVIEW

File: docs/decisions/0021-transaction-aggregates.rst [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+Inter-service Communication
+---------------------------
+There are several options for how transaction/reversal data can make its way to the aggregates
+layer in the enterprise-access service:
+
+1. Via management commands that reads transaction and reversal records via the enterprise-subsidy
+   REST API.  This approach may work for backpopulation, but is generally disfavorable because
+   it won't help us keep the aggregate records up-to-date in a timely manner (we'd have to run
+   it on some cron/schedule).
+2. Updating the aggregates based on the ``redeem`` response body (on successful redemption).
```
Comment (macdiesel) [inline docs/decisions/0021-transaction-aggregates.rst:66]: nit: If we will not use option 2 or 3 should they be moved to the "Alternatives Considered" block?

File: docs/decisions/0021-transaction-aggregates.rst [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+Additionally, we may eventually want aggregates at the **(policy, learner, group)** level, to support
+business record configurations where multiple groups relate to a single policy.
+
+Once these aggregates are in place, replace requests to the enterpirse-subsidy service
+in the ``can_redeem`` and ``redeem`` flows with database queries against these aggregates.
+
+Maintaining these aggregates depends on sourcing transaction and reversal data as follows:
+
+1. Starting from a zero-state in which Aggregate models exist but are not yet populated (backpopulation).
```
Comment (macdiesel) [inline docs/decisions/0021-transaction-aggregates.rst:53]: I think we should also call out recalculation here too.

File: docs/decisions/0021-transaction-aggregates.rst [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+3. Emiting transaction/reversal commit events from the enterprise-subsidy service. If we can
+   emit these to an event broker, consumable by enterprise-access, then we can update aggregates
+   as transactions and reversals are committed. Furthermore, we could utilize a Django management
+   command to emit events that already exist (perhaps with ``modified`` times
+   earlier than some adjustable value) as a way to replay history. This would allow us to backpopulate
+   aggregate records.
+
+We'll utilize option (3) to both backpopulate and update aggregate records in real time.
+
+Ensuring Data Consistency
```
Comment (macdiesel) [inline docs/decisions/0021-transaction-aggregates.rst:79]: I'm operating under the assumption that aggregations are calculated initially from a zero state and are thereafter cumulative. 

If this is the case I think we will need some type of job to monitor the aggregation and ensure they are correct. Loops back to my earlier comment that we will need a way to recalculate as well.

File: docs/decisions/0021-transaction-aggregates.rst [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+If we replace the synchronous querying of the enterprise-subsidy service for aggregates, with
+database reads of aggregate tables populated by asynchronous events, we've introduced asynchronicity
+into our overall flow.  This introduces the risk of running ``can_redeem`` flows with transaction
+aggregates that are stale, i.e. individual transaction records exist that have not been
+taken into accounts in the aggregate records. Without mitigation, this risks that our overall
+``redeem`` flow becomes less consistent, and timely consistency is one of the key cross-functional
+requirements of learner credit redemption.
+
+One possible mitigation is to record the timestamp of the most recent transaction or reversal
```
Comment (macdiesel) [inline docs/decisions/0021-transaction-aggregates.rst:104]: +1 solid idea.

Thinking out loud I wonder what other ways we could go about check summing this. This approach would ensure we are pointing at the same index but it doesn't ensure we've read all the records... I'm not sure, maybe the offset is enough.

**Review by macdiesel — APPROVED** [review]
Seems reasonable. I left some thoughts. I do think that there could be a performance win here, might be worth calling that out more prominently.

File: docs/decisions/0021-transaction-aggregates.rst (hunk 1 above)
Comment (iloveagent57) [reply docs/decisions/0021-transaction-aggregates.rst:66]: I view "Alternatives considered" as a place to document different approaches to the general problem.  This is more like the rejected alternatives to a sub-problem: having accepted that we'll maintain aggregates in enterprise-access, how will the data get there?

File: docs/decisions/0021-transaction-aggregates.rst (hunk 2 above)
Comment (iloveagent57) [reply docs/decisions/0021-transaction-aggregates.rst:53]: Good call, added a note about this, specifically for use in error correction.

File: docs/decisions/0021-transaction-aggregates.rst (hunk 3 above)
Comment (iloveagent57) [reply docs/decisions/0021-transaction-aggregates.rst:79]: It's a good idea, added it as another possible mitigation.

File: docs/decisions/0021-transaction-aggregates.rst (hunk 4 above)
Comment (iloveagent57) [reply docs/decisions/0021-transaction-aggregates.rst:104]: Yeah, it's a good point. I think there's one more thing we have to take into account, which is the absence of a needed aggregate for some `redeem` flow - if you query for aggregates on a given (policy, learner), and no record is in the DB, do we assume the metric is 0, or that we haven't yet computed the aggregate? I think we can hook into the `can-redeem` flow, plus maybe formalize a `zero aggregate` concept to mitigate this. I'll add a paragraph to the doc.

File: docs/decisions/0021-transaction-aggregates.rst (hunk 1 above)
Comment (macdiesel) [reply docs/decisions/0021-transaction-aggregates.rst:68]: No strong feelings.

**Pull request merged**